File: lab/may_16/solution.py
# author: jamie
# email: jinjiedeng.jjd@gmail.com

if __name__ == "__main__":
    n, k = map(int, input().split())
    S = list(map(int, input().split()))
    D = {}
    U = []
    for i in range(k):
        if D.get(S[i]) is None: D[S[i]] = 0
        D[S[i]] += 1
    U.append(len(D))
    for i in range(k, n):
        D[S[i - k]] -= 1
        if D[S[i - k]] == 0: D.pop(S[i - k])
        if D.get(S[i]) is None: D[S[i]] = 0
        D[S[i]] += 1
        U.append(len(D))
    print(" ".join(map(str, U)))

File: lab/may_16/solution.cc
/*
* author: jamie
* email: jinjiedeng.jjd@gmail.com
*/
#include <iostream>
#include <map>
#include <vector>

int main ()
{
  uint64_t n, k;
  std::cin >> n >> k;
  std::vector<uint64_t> S(n);
  std::map<uint64_t, uint64_t> map;
  for (auto& s : S) { std::cin >> s; }
  for (uint64_t i {}; i != k; ++i) { ++map[S[i]]; }
  std::cout << map.size();
  for (uint64_t i {k}; i != n; ++i)
  {
    if (--map[S[i - k]] == 0) { map.erase(S[i - k]); }
    ++map[S[i]];
    std::cout << " " << map.size();
  }
  return 0;
}